src/java.base/share/classes/java/util/concurrent/ScheduledThreadPoolExecutor.java

```diff
@@ -525,25 +530,7 @@ private long triggerTime(long delay, TimeUnit unit) {
      * Returns the nanoTime-based trigger time of a delayed action.
      */
     long triggerTime(long delay) {
-        return System.nanoTime() +
-            ((delay < (Long.MAX_VALUE >> 1)) ? delay : overflowFree(delay));
-    }
-
-    /**
-     * Constrains the values of all delays in the queue to be within
-     * Long.MAX_VALUE of each other, to avoid overflow in compareTo.
-     * This may occur if a task is eligible to be dequeued, but has
-     * not yet been, while some other task is added with a delay of
-     * Long.MAX_VALUE.
-     */
-    private long overflowFree(long delay) {
-        Delayed head = (Delayed) super.getQueue().peek();
-        if (head != null) {
-            long headDelay = head.getDelay(NANOSECONDS);
-            if (headDelay < 0 && (delay - headDelay < 0))
-                delay = Long.MAX_VALUE + headDelay;
-        }
-        return delay;
+        return System.nanoTime() + Math.min(delay, MAX_NANOS);
```

Comment:
The hi-res time source could have any long value so I'm wondering if it's okay for triggerTime to be negative.

Reply:
System.nanoTime can be negative anyway (and only checked by subtraction). So this should be OK.